Basic core dev docs (#2732)
casper-network/casper-node: book/src/architecture/announcement-howto.md

# HOWTO: Adding an announcement

A question that sometimes comes up is how to notify a component about a change that happened in another one. As components are decoupled they do not know about specific other components, so the way we achieve this is by having one component announce new information, and one or more react to said announcement, all wired up inside the reactor.

> Let's look at a concrete example (that we will never implement): We want the diagnostics console to send a message to all connected clients whenever a node starts or ceases to be come a validator. Our plan is to

Reviewer: ```suggestion
Let's look at a concrete example (that we will never implement): We want the diagnostics console to send a message to all connected clients whenever a node starts or ceases to become a validator. Our plan is to
```

This is usually simple to do, since after a brief start, the compiler will guide us through most steps necessary.

## Adding an announcement

First, we create a new announcement in `announcements.rs`.

```rust,noplayground
/// An announcement indicating that the validator/non-validator status of the node changed.
#[derive(Debug, Serialize)]
pub(crate) struct ValidatorStatusChangedAnnouncement {
    /// Whether or not we are a validator now.
    pub(crate) new_status: bool,
}
```

A `Display` implementation will be needed later:

```rust,noplayground
impl Display for ValidatorStatusChangedAnnouncement {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "validator status changed to {}", self.new_status)
    }
}
```

We also add a method in the `EffectBuilder` in `effects.rs` to emit it:

```rust,noplayground
/// Announce that the validator status has been changed.
pub(crate) async fn announce_validator_status_changed(self, new_status: bool)
where
    REv: From<ValidatorStatusChangedAnnouncement>,
{
    self.event_queue
        .schedule(ValidatorStatusChangedAnnouncement { new_status }, QueueKind::Regular)
        .await
}
```

> Now we can call `effect_builder.announce_validator_status_changed(..)` in another component and use it to return an effect that will cause the announcement to emitted.

Reviewer: ```suggestion
Now we can call `effect_builder.announce_validator_status_changed(..)` in another component and use it to return an effect that will cause the announcement to be emitted.
```

```

This tells us that the participating reactor does not handle the `ValidatorStatusChangedAnnouncement`, but since it has a consensus component, it must do so. The only recommended way of handling any kind of event is by adding a variant for it to the reactor component, for which we use the `#[from]` macro:

```rust,noplayground
pub(crate) enum ParticipatingEvent {
    // ...
    /// Validator status change announcement.
    #[from]
    ValidatorStatusChangedAnnouncement(ValidatorStatusChangedAnnouncement),
```

The compiler will tell us that we need to handle that variant in the routing code

```
error[E0004]: non-exhaustive patterns: `ValidatorStatusChangedAnnouncement(_)` not covered
   --> node/src/reactor/participating.rs:776:15
```

so we add an action, by just logging it -- we do not have an interested party for it yet:

```rust,noplayground
fn dispatch_event(
  // ...
    ParticipatingEvent::ValidatorStatusChangedAnnouncement(_) => {
        debug!("TODO: swallowed a validator status changed announcement");
        Effects::new()
    }
  // ...
```

The necessary additions to the `description` and `Display::fmt` methods of `ParticipatingEvent` are not shown here.

> With these changes, the node will actually compile again and output a debug-level log messages when the announcement is made. We can now continue by implementing a consumer of this announcement.

Reviewer: ```suggestion
With these changes, the node will actually compile again and output debug-level log messages when the announcement is made. We can now continue by implementing a consumer of this announcement.
```
